File: backend/app/schemas/team.py
"""Pydantic schemas for teams."""

import re
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field, validator


def slugify(name: str) -> str:
    """Convert name to URL-safe slug."""
    slug = name.lower().strip()
    slug = re.sub(r"[^\w\s-]", "", slug)
    slug = re.sub(r"[\s_-]+", "-", slug)
    slug = re.sub(r"^-+|-+$", "", slug)
    return slug


class TeamBase(BaseModel):
    """Base team schema."""

    name: str = Field(..., min_length=1, max_length=255, description="Team name")
    description: str | None = Field(None, description="Team description")
    slug: str | None = Field(None, min_length=1, max_length=100, description="URL-safe slug")
    metadata: dict[str, Any] | None = Field(default_factory=dict)

    @validator("slug", pre=True, always=True)
    def generate_slug(cls, v, values):
        """Auto-generate slug from name if not provided."""
        if v:
            return v
        if "name" in values:
            return slugify(values["name"])
        return None


class CreateTeamRequest(TeamBase):
    """Request schema for creating a team."""

    domain_id: UUID = Field(..., description="Domain ID this team belongs to")


class UpdateTeamRequest(BaseModel):
    """Request schema for updating a team."""

    name: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None
    slug: str | None = Field(None, min_length=1, max_length=100)
    is_active: bool | None = None
    metadata: dict[str, Any] | None = None


class TeamMemberBase(BaseModel):
    """Base team member schema."""

    user_id: UUID
    role: str = Field(default="member", description="Role in team: team-lead or member")


class AddTeamMemberRequest(TeamMemberBase):
    """Request schema for adding a member to team."""

    pass


class UpdateTeamMemberRequest(BaseModel):
    """Request schema for updating team member role."""

    role: str = Field(..., description="Role in team: team-lead or member")


class TeamMemberResponse(TeamMemberBase):
    """Response schema for team member."""

    email: str | None = None
    full_name: str | None = None
    joined_at: datetime

    class Config:
        from_attributes = True


class TeamResponse(TeamBase):
    """Response schema for team."""

    id: UUID
    domain_id: UUID
    workspace_id: UUID
    is_active: bool
    created_by: UUID | None
    created_at: datetime
    updated_at: datetime
    members_count: int | None = Field(None, description="Number of members in team")

    class Config:
        from_attributes = True


class TeamWithMembers(TeamResponse):
    """Team response with members included."""

    members: list[TeamMemberResponse] = Field(default_factory=list, description="Team members")

    class Config:
        from_attributes = True


class TeamHierarchyResponse(BaseModel):
    """Hierarchical team structure response."""

    workspace_id: UUID
    organization_name: str
    domains: list[dict[str, Any]] = Field(default_factory=list)

    class Config:
        from_attributes = True

File: frontend/src/services/team.ts
/**
 * Team API service
 */
import { api } from './api';

export interface TeamMember {
  user_id: string;
  email?: string;
  full_name?: string;
  role: string;
  joined_at: string;
}

export interface Team {
  id: string;
  domain_id: string;
  workspace_id: string;
  name: string;
  description?: string;
  slug: string;
  is_active: boolean;
  metadata?: Record<string, any>;
  created_by?: string;
  created_at: string;
  updated_at: string;
  members_count?: number;
  members?: TeamMember[];
}

export interface Domain {
  id: string;
  workspace_id: string;
  name: string;
  description?: string;
  slug: string;
  is_active: boolean;
  metadata?: Record<string, any>;
  created_by?: string;
  created_at: string;
  updated_at: string;
  teams_count?: number;
  teams?: Team[];
}

export interface TeamHierarchy {
  workspace_id: string;
  workspace_name: string;
  domains: Array<{
    id: string;
    name: string;
    slug: string;
    description?: string;
    teams: Array<{
      id: string;
      name: string;
      slug: string;
      description?: string;
      is_active: boolean;
    }>;
  }>;
}

export interface CreateDomainRequest {
  name: string;
  description?: string;
  slug?: string;
  metadata?: Record<string, any>;
}

export interface UpdateDomainRequest {
  name?: string;
  description?: string;
  slug?: string;
  is_active?: boolean;
  metadata?: Record<string, any>;
}

export interface CreateTeamRequest {
  domain_id: string;
  name: string;
  description?: string;
  slug?: string;
  metadata?: Record<string, any>;
}

export interface UpdateTeamRequest {
  name?: string;
  description?: string;
  slug?: string;
  is_active?: boolean;
  metadata?: Record<string, any>;
}

export interface AddTeamMemberRequest {
  user_id: string;
  role: string;
}

export interface UpdateTeamMemberRequest {
  role: string;
}

class TeamService {
  // Domain endpoints
  async createDomain(orgId: string, data: CreateDomainRequest): Promise<Domain> {
    const response = await api.post(`/workspaces/${orgId}/domains`, data);
    return response.data;
  }

  async getDomains(orgId: string): Promise<Domain[]> {
    const response = await api.get(`/workspaces/${orgId}/domains`);
    return response.data;
  }

  async getDomain(orgId: string, domainId: string): Promise<Domain> {
    const response = await api.get(`/workspaces/${orgId}/domains/${domainId}`);
    return response.data;
  }

  async updateDomain(orgId: string, domainId: string, data: UpdateDomainRequest): Promise<Domain> {
    const response = await api.patch(`/workspaces/${orgId}/domains/${domainId}`, data);
    return response.data;
  }

  async deleteDomain(orgId: string, domainId: string): Promise<void> {
    await api.delete(`/workspaces/${orgId}/domains/${domainId}`);
  }

  // Team endpoints
  async createTeam(orgId: string, data: CreateTeamRequest): Promise<Team> {
    const response = await api.post(`/workspaces/${orgId}/teams`, data);
    return response.data;
  }

  async getTeams(orgId: string, domainId?: string): Promise<Team[]> {
    const params = domainId ? { domain_id: domainId } : {};
    const response = await api.get(`/workspaces/${orgId}/teams`, { params });
    return response.data;
  }

  async getTeam(orgId: string, teamId: string): Promise<Team> {
    const response = await api.get(`/workspaces/${orgId}/teams/${teamId}`);
    return response.data;
  }

  async updateTeam(orgId: string, teamId: string, data: UpdateTeamRequest): Promise<Team> {
    const response = await api.patch(`/workspaces/${orgId}/teams/${teamId}`, data);
    return response.data;
  }

  async deleteTeam(orgId: string, teamId: string): Promise<void> {
    await api.delete(`/workspaces/${orgId}/teams/${teamId}`);
  }

  async getHierarchy(orgId: string): Promise<TeamHierarchy> {
    const response = await api.get(`/workspaces/${orgId}/teams/hierarchy`);
    return response.data;
  }

  // Team member endpoints
  async addMember(orgId: string, teamId: string, data: AddTeamMemberRequest): Promise<void> {
    await api.post(`/workspaces/${orgId}/teams/${teamId}/members`, data);
  }

  async getMembers(orgId: string, teamId: string): Promise<TeamMember[]> {
    const response = await api.get(`/workspaces/${orgId}/teams/${teamId}/members`);
    return response.data;
  }

  async updateMemberRole(orgId: string, teamId: string, userId: string, data: UpdateTeamMemberRequest): Promise<void> {
    await api.patch(`/workspaces/${orgId}/teams/${teamId}/members/${userId}`, data);
  }

  async removeMember(orgId: string, teamId: string, userId: string): Promise<void> {
    await api.delete(`/workspaces/${orgId}/teams/${teamId}/members/${userId}`);
  }
}

export default new TeamService();
